src/persist-types/src/arrow.rs

```diff
@@ -0,0 +1,211 @@
+            proto::data_type::Kind::Null(()) => DataType::Null,
+            proto::data_type::Kind::Boolean(()) => DataType::Boolean,
+            proto::data_type::Kind::Uint8(()) => DataType::UInt8,
+            proto::data_type::Kind::Uint16(()) => DataType::UInt16,
+            proto::data_type::Kind::Uint32(()) => DataType::UInt32,
+            proto::data_type::Kind::Uint64(()) => DataType::UInt64,
+            proto::data_type::Kind::Int8(()) => DataType::Int8,
+            proto::data_type::Kind::Int16(()) => DataType::Int16,
+            proto::data_type::Kind::Int32(()) => DataType::Int32,
+            proto::data_type::Kind::Int64(()) => DataType::Int64,
+            proto::data_type::Kind::Float32(()) => DataType::Float32,
+            proto::data_type::Kind::Float64(()) => DataType::Float64,
+            proto::data_type::Kind::String(()) => DataType::Utf8,
+            proto::data_type::Kind::Binary(()) => DataType::Binary,
+            proto::data_type::Kind::FixedBinary(size) => DataType::FixedSizeBinary(size),
+            proto::data_type::Kind::List(inner) => DataType::List(Arc::new((*inner).into_rust()?)),
+            proto::data_type::Kind::Map(inner) => {
+                let value = inner
+                    .value
+                    .ok_or_else(|| TryFromProtoError::missing_field("map.value"))?;
+                DataType::Map(Arc::new((*value).into_rust()?), inner.sorted)
+            }
+            proto::data_type::Kind::Struct(inner) => {
+                let children: Vec<Field> = inner
+                    .children
+                    .into_iter()
+                    .map(|c| c.into_rust())
+                    .collect::<Result<_, _>>()?;
+                DataType::Struct(Fields::from(children))
+            }
+        };
+
+        Ok(data_type)
+    }
+}
+
+impl RustType<proto::Field> for arrow::datatypes::Field {
+    fn into_proto(&self) -> proto::Field {
+        proto::Field {
+            name: self.name().clone(),
+            nullable: self.is_nullable(),
+            data_type: Some(Box::new(self.data_type().into_proto())),
+        }
+    }
+
+    fn from_proto(proto: proto::Field) -> Result<Self, TryFromProtoError> {
+        let proto::Field {
+            name,
+            nullable,
+            data_type,
+        } = proto;
+        let data_type =
+            data_type.ok_or_else(|| TryFromProtoError::missing_field("field.data_type"))?;
+        let data_type = (*data_type).into_rust()?;
+
+        Ok(Field::new(name, data_type, nullable))
+    }
+}
+
+impl RustType<proto::Buffer> for arrow::buffer::Buffer {
+    fn into_proto(&self) -> proto::Buffer {
+        // TODO(parkmycar): There is probably something better we can do here.
+        proto::Buffer {
+            data: bytes::Bytes::from(self.as_slice().to_vec()),
+        }
+    }
+
+    fn from_proto(proto: proto::Buffer) -> Result<Self, TryFromProtoError> {
+        Ok(arrow::buffer::Buffer::from_bytes(proto.data.into()))
```

Comment:
What's this `into()`? Isn't `data` already a `Bytes`?

Reply:
It turns out arrow has its own [`Bytes`](https://github.com/apache/arrow-rs/blob/master/arrow-buffer/src/bytes.rs) type that has an `O(1)` conversion from `bytes::Bytes`